File: extension/src/score/embeddingBackend.ts
import { lexiconBytes, parseLexicon, type Lexicon } from "./lexicon";
import artifact from "./lexicon.json";
import { EMBEDDING_DIMENSIONS, embedText, normalizeVector, tokenize } from "./textEmbedding";

export const LEXICON_ARTIFACT_VERSION = 1;

export const HASHED_TERMS_IDENTITY = "hashed-terms";

export interface EmbeddingBackend {
  readonly identity: string;
  readonly dimensions: number;
  embed(text: string): number[] | null;
}

export function hashedTerms(dimensions = EMBEDDING_DIMENSIONS): EmbeddingBackend {
  return {
    identity: `${HASHED_TERMS_IDENTITY}/${dimensions}`,
    dimensions,
    embed: (text) => embedText(text, dimensions),
  };
}

export function fromLexicon(lexicon: Lexicon, identity: string): EmbeddingBackend {
  return {
    identity,
    dimensions: lexicon.dimensions,
    embed(text: string): number[] | null {
      const total = new Array<number>(lexicon.dimensions).fill(0);
      let found = 0;
      for (const token of tokenize(text)) {
        const row = lexicon.row(token);
        if (row === null) {
          continue;
        }
        found++;
        for (let i = 0; i < lexicon.dimensions; i++) {
          total[i] = (total[i] as number) + (row[i] as number);
        }
      }
      if (found === 0) {
        return null;
      }
      return normalizeVector(total);
    },
  };
}

// an absent or unreadable table is a table this build does not have, and a build without one
// still has to score, so the caller is handed the hashed terms rather than nothing
export function backendFromArtifact(data: unknown, fallback = hashedTerms()): EmbeddingBackend {
  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    return fallback;
  }
  const record = data as Record<string, unknown>;
  if (record.artifactVersion !== LEXICON_ARTIFACT_VERSION || record.present !== true) {
    return fallback;
  }
  const { identity, bytes } = record;
  if (typeof identity !== "string" || identity.length === 0 || typeof bytes !== "string") {
    return fallback;
  }
  const decoded = lexiconBytes(bytes);
  if (decoded === null) {
    return fallback;
  }
  const lexicon = parseLexicon(decoded);
  if (lexicon === null) {
    return fallback;
  }
  return fromLexicon(lexicon, identity);
}

let bundled: EmbeddingBackend | null = null;

// decoding the table costs megabytes of work, so it happens once per document
export function bundledEmbeddingBackend(): EmbeddingBackend {
  bundled ??= backendFromArtifact(artifact);
  return bundled;
}

File: research/verdict_research/features/embedding_backend.py
import base64
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Protocol

from verdict_research.features.lexicon import Lexicon, parse_lexicon
from verdict_research.features.text_embedding import (
    EMBEDDING_DIMENSIONS,
    embed_text,
    normalize_vector,
    tokenize,
)

LEXICON_ARTIFACT_VERSION = 1

HASHED_TERMS_IDENTITY = "hashed-terms"


class EmbeddingBackend(Protocol):
    identity: str
    dimensions: int

    def embed(self, text: str) -> list[float] | None: ...


@dataclass
class HashedTermsBackend:
    dimensions: int = EMBEDDING_DIMENSIONS

    @property
    def identity(self) -> str:
        return f"{HASHED_TERMS_IDENTITY}/{self.dimensions}"

    def embed(self, text: str) -> list[float] | None:
        return embed_text(text, self.dimensions)


@dataclass
class LexiconBackend:
    lexicon: Lexicon
    identity: str

    @property
    def dimensions(self) -> int:
        return self.lexicon.dimensions

    def embed(self, text: str) -> list[float] | None:
        total = [0.0] * self.lexicon.dimensions
        found = 0
        for token in tokenize(text):
            row = self.lexicon.row(token)
            if row is None:
                continue
            found += 1
            for i in range(self.lexicon.dimensions):
                total[i] += row[i]
        if found == 0:
            return None
        return normalize_vector(total)


def hashed_terms(dimensions: int = EMBEDDING_DIMENSIONS) -> HashedTermsBackend:
    return HashedTermsBackend(dimensions=dimensions)


def from_lexicon(lexicon: Lexicon, identity: str) -> LexiconBackend:
    return LexiconBackend(lexicon=lexicon, identity=identity)


def backend_from_artifact(
    data: object, fallback: EmbeddingBackend | None = None
) -> EmbeddingBackend:
    resolved = hashed_terms() if fallback is None else fallback
    if not isinstance(data, dict):
        return resolved
    if data.get("artifactVersion") != LEXICON_ARTIFACT_VERSION or data.get("present") is not True:
        return resolved
    identity = data.get("identity")
    encoded = data.get("bytes")
    if not isinstance(identity, str) or not identity or not isinstance(encoded, str):
        return resolved
    try:
        decoded = base64.b64decode(encoded, validate=True)
    except (ValueError, TypeError):
        return resolved
    lexicon = parse_lexicon(decoded)
    if lexicon is None:
        return resolved
    return from_lexicon(lexicon, identity)


# the same file the extension bundles, because a corpus featurised with one embedding and
# scored with another is a corpus of numbers that mean nothing
BUNDLED_ARTIFACT = (
    Path(__file__).resolve().parents[3] / "extension" / "src" / "score" / "lexicon.json"
)

_bundled: EmbeddingBackend | None = None


def _read(path: Path) -> EmbeddingBackend:
    try:
        return backend_from_artifact(json.loads(path.read_text(encoding="utf-8")))
    except (OSError, json.JSONDecodeError):
        return hashed_terms()


def bundled_embedding_backend(path: Path | None = None) -> EmbeddingBackend:
    global _bundled
    if path is not None:
        return _read(path)
    if _bundled is None:
        _bundled = _read(BUNDLED_ARTIFACT)
    return _bundled
